ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -528,15 +543,47 @@ RouterException getChunkException() {
      *                                    created as part of this poll operation.
      */
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
-      //First, check if any of the existing requests have timed out.
-      cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!maybeProcessDecryptCallbackAndCompleteOperation()) {
+        //First, check if any of the existing requests have timed out.
+        cleanupExpiredInFlightRequests();
+        checkAndMaybeComplete();
+        if (!isComplete()) {
+          fetchRequests(requestRegistrationCallback);
+        }
       }
     }
 
     /**
+     * Maybe process decrypt callback and complete the operation if applicable. This is a no-op for non-encrypted blobs
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    protected boolean maybeProcessDecryptCallbackAndCompleteOperation() {
+      if (decryptionRequired.get() && decryptCallbackResultInfo.resultAvailable) {
+        logger.trace("Processing decrypt callback stored result for data chunk {}", chunkBlobId);
+        if (decryptCallbackResultInfo.exception == null && (getChunkException() == null
+            || !retainChunkExceptionOnSuccess)) {
+          chunkIndexToBuffer.put(chunkIndex,
+              filterChunkToRange(decryptCallbackResultInfo.result.getDecryptedBlobContent()));
+          numChunksRetrieved++;
+          logger.trace("Decrypt result successfully updated for data chunk {}", chunkBlobId);
+          decryptionStatusTracker.setSucceeded();
+        } else {
+          logger.trace("Setting operation exception as decryption callback invoked with exception {} for data chunk {}",
+              decryptCallbackResultInfo.exception, chunkBlobId);
```

Comment:
I think all this can be set as part of the decryption call back (similar to my comment for `GetBlobInfo`).

Reply:
I don't think we can do that. Will sync up with you offline on this.